### ISSUE TITLE
Add support for reading dependencies

### ISSUE THREAD

**Issue by jaraco** [python/importlib_metadata#18]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Sep 26, 2018, 24:57

I recently stumbled onto another case where `pkg_resources` supplies some important metadata, the dependencies for a package.

```
~ $ pkg_resources.get_distribution('jaraco.functools').requires()
[Requirement.parse('more-itertools')]
```

I believe importlib_metadata should probably provide at least a list of requirements if not also possibly parsing those requirements (although the latter might only work if the third-party 'packaging' library is available).

**Comment by jaraco** [comment]
In GitLab by [@warsaw](https://gitlab.com/warsaw) on Sep 28, 2018, 22:14

Is this just a matter of parsing the `Requires-Dist` headers from the `METADATA` file?  Would the API have to mirror `pkg_resources`?

+1 for providing that as a convenience.

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Nov 11, 2018, 02:03

> Is this just a matter of parsing the `Requires-Dist` headers from the `METADATA` file?

Yes, probably.

> Would the API have to mirror `pkg_resources`?

Not necessarily.

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Nov 11, 2018, 02:03

assigned to @jaraco

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Nov 11, 2018, 02:11

> Is this just a matter of parsing the `Requires-Dist` headers from the `METADATA` file?

In order to support editable installs, it would also need to parse out the `egg-info/requires.txt`.

It should also probably present which 'extras' are provided and what their dependencies are...

And it should either wrap the results in objects from `packaging` or provide instructions on how to use the outputs from to pass to `packaging` for enhanced functionality.

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Nov 11, 2018, 02:31

Gah. The `egg-info` format for requires.txt looks like this:

```
dep1
dep2

[extra]
extra1dep1

[extra2]
extra2dep1
```

That file, although it looks like an .ini file doesn't have a heading for the first section, so fails when loaded with a ConfigParser.

Looking at pkg_resources, it uses [this function](https://github.com/pypa/setuptools/blob/7f310da972ee9d979b8b0bfb9c19cabe91bf94a2/pkg_resources/__init__.py#L3064) to hand-parse the file. I'm wondering if it's safe to just inject a special section and then use the configparser.

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Nov 11, 2018, 02:35

Nope. ConfigParser also doesn't read simple lines from sections.

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Nov 11, 2018, 02:59

mentioned in commit 0ab7f13cd79e6b26839b77471fe1d5cf89773c37

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Nov 11, 2018, 03:01

This work is once again making me want to revisit the idea of splitting PathDistribution into separate classes for dist-info and egg-info. It's also making me wonder if we shouldn't instead concentrate our efforts on eliminating egg-info and getting proper editable installs into pip that generate dist-info. I know there are some efforts under consideration to make that happen.

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Dec 1, 2018, 02:56

mentioned in commit 43185e059d899a0abda01e95b04f32d61f614f32

**Comment by jaraco** [comment]
In GitLab by [@warsaw](https://gitlab.com/warsaw) on Dec 3, 2018, 19:31

@jaraco Yes, ideally we'd be able to eliminate egg-infos, but gosh that does seem like long-tail territory.  Given that at least for now we have to support editable installs, it seems like that forces our hand for egg-info support.

If you split `PathDistribution` will you also have to split `WheelDistribution` too?

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Dec 7, 2018, 22:18

> If you split `PathDistribution` will you also have to split `WheelDistribution` too?

No, because wheels will never be eggs. But if we decide to support eggs, that will have a comparable split.

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Dec 16, 2018, 19:45

mentioned in merge request !31

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Dec 23, 2018, 19:05

mentioned in commit 912ad777ba3fe28618454c240f296303482dcbcc

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Dec 23, 2018, 19:07

mentioned in merge request !34

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Dec 23, 2018, 19:08

mentioned in commit bc2be67fd27307e4337f65c8f8434f98a9fbfa13

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Dec 30, 2018, 16:46

closed via merge request !34

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Dec 30, 2018, 16:46

closed via commit bc2be67fd27307e4337f65c8f8434f98a9fbfa13
